### PR TITLE
Support cover_all options

### PR DESCRIPTION
**Pull request by rezoo**
This PR supports `cover_all` options in `MaxPooling2D`, and `MaxPoolingND`.
- [x] Support pooling functions
- [x] Write tests for pooling functions

### CODE REVIEW

**Comment by rezoo** [comment]
I finished implementation. Please check it.

File: onnx_chainer/functions/pooling.py [hunk 1]
```diff
@@ -3,24 +3,28 @@
...
             strides=(func.sy, func.sx)
         ),
 
 
 def convert_AveragePoolingND(func, onnx_op_name, opset_version, input_names, output_names, parameters):
-    pad = []
-    x_ndim = len(func.inputs[0].shape)
-    k_ndim = len(func.ksize)
-    for _ in range(x_ndim - k_ndim):
-        pad.append(0)
```
Comment (mitmul) [inline onnx_chainer/functions/pooling.py]: Why did you remove this part? I think Chainer doesn't have a way to pad along batch dimension or channel dimension, but the ONNX operator `AveragePool` can represent such case. So I just filled with the dimensions other than the spatial dimensions.

File: tests/functions_tests/test_poolings.py [hunk 2]
```diff
@@ -10,31 +10,32 @@
...
-     'in_shape': (1, 3, 6, 6, 6), 'args': [2, 1, 0]},
-    {'name': 'max_pooling_2d', 'ops': F.max_pooling_2d,
-     'in_shape': (1, 3, 6, 6), 'args': [2, 1, 0]},
-    {'name': 'max_pooling_nd', 'ops': F.max_pooling_nd,
-     'in_shape': (1, 3, 6, 6, 6), 'args': [2, 1, 0]},
-)
+@testing.parameterize(*(testing.product({
+    'name': ['max_pooling_2d'],
+    'in_shape': [(1, 3, 6, 5)],
+    'args': [[2, 2, 0]],
```
Comment (mitmul) [inline tests/functions_tests/test_poolings.py]: Is there any particular reason why you changed the arguments in addition to adding `cover_all` option to test?

File: onnx_chainer/functions/pooling.py (hunk 1 above)
Comment (rezoo) [reply onnx_chainer/functions/pooling.py]: > but the ONNX operator AveragePool can represent such case.

? Sorry, but at least it seems that `AveragePool` can accept `pad` without filling the batch and the channel dimensions, and as far as we refer to the examples of AveragePool [1], the approach without batch and channel is valid.

[1] https://github.com/onnx/onnx/blob/master/docs/Operators.md#AveragePool

File: tests/functions_tests/test_poolings.py (hunk 2 above)
Comment (mitmul) [reply tests/functions_tests/test_poolings.py]: OK, thanks!

File: onnx_chainer/functions/pooling.py (hunk 1 above)
Comment (mitmul) [reply onnx_chainer/functions/pooling.py]: Oops, sorry I was wrongly taking the non-spatial dimensions into account when converting pooling functions... Now it's clear!

**Comment by rezoo** [comment]
Hello, I'd be glad if you could tell any progress.

File: onnx_chainer/functions/pooling.py [hunk 3]
```diff
@@ -3,18 +3,28 @@
...
+    else:
+        dph, dpw = 0, 0
     if opset_version == 1 or opset_version == 7:
         return helper.make_node(
             onnx_op_name, input_names, output_names,
             kernel_shape=(func.kh, func.kw),
-            pads=(func.ph, func.pw, func.ph, func.pw),
-            strides=(func.sy, func.sx),
-            count_include_pad=1,
+            pads=(func.ph, func.pw, func.ph + dph, func.pw + dpw),
```
Comment (mitmul) [inline onnx_chainer/functions/pooling.py]: @rezoo Sorry for the delay. When a Function which has `ksize=2`, `stride=2`, and `pad=1` with `cover_all=True` is given to this converter, `dph` and `dpw` will both be 1, and then `pads` for `make_node` function will be `(1, 1, 2, 2)`. This causes “Pad should be smaller than kernel” error when the converted ONNX operator is run with ONNXRuntime. This specification seems not to be written in the official document of ONNX, but you can confirm this by adding `pad=1` cases to the current tests.

File: tests/functions_tests/test_poolings.py [hunk 4]
```diff
@@ -10,31 +10,32 @@
...
+    'name': ['max_pooling_nd'],
+    'in_shape': [(1, 3, 6, 5, 4)],
+    'args': [[2, 2, 0]],
+    'cover_all': [False, True],
+}) + testing.product({
+    'name': ['average_pooling_nd'],
+    'in_shape': [(1, 3, 6, 5, 4)],
+    'args': [[2, 2, 0]],
+    'cover_all': [None],
+})))
```
Comment (mitmul) [inline tests/functions_tests/test_poolings.py]: Could you give `pad=1` to these test cases instead of `pad=0`?
And could you add another test case to check the behavior when stride > ksize? like:
```
testing.product({
    'name': ['max_pooling_2d'],
    'in_shape': [(1, 3, 6, 5)],
    'args': [[2, 3, 1]],
    'cover_all': [True],
})))
```
and please raise a runtime error when stride larger than ksize is given to the converters.

**Comment by rezoo** [comment]
I fixed the points you have mentioned. Can you check them?

**Comment by mitmul** [comment]
Thank you for updating. LGTM
(I’ll add some tests to ensure the errors will correctly be raised)
